server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -668,51 +668,57 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
+                            // The commit failed (i.e. master is failing over) but this does not imply that the join has actually failed:
+                            // the next master may have already accepted the state that we just published and will therefore include the
+                            // joining node in its future states too. Thus we need to wait for the next committed state before we know the
+                            // eventual outcome, and we need to wait for that before we can release (our ref to) the connection and complete
+                            // the listener.
+
+                            // NB we are on the master update thread here at the end of processing the failed cluster state update, so this
+                            // all happens before any cluster state update that re-elects a master
+                            assert ThreadPool.assertCurrentThreadPool(MasterService.MASTER_UPDATE_THREAD_NAME);
+
+                            final ClusterStateListener clusterStateListener = new ClusterStateListener() {
+                                @Override
+                                public void clusterChanged(ClusterChangedEvent event) {
+                                    final var discoveryNodes = event.state().nodes();
+                                    // Keep the connection open until the next committed state
+                                    if (discoveryNodes.getMasterNode() != null) {
+                                        // Remove this listener to avoid memory leaks
+                                        clusterService.removeListener(this);
+
+                                        if (discoveryNodes.nodeExists(joinRequest.getSourceNode().getId())) {
+                                            ll.onResponse(null);
+                                        } else {
+                                            ll.onFailure(e);
                                         }
                                     }
-                                };
-
-                                clusterService.addListener(clusterStateListener);
-                            } else {
-                                Releasables.close(response);
-                                joinListener.onFailure(e);
+                                }
+                            };
+                            clusterService.addListener(clusterStateListener);
+
+                            // Another node was elected, and doesn't have the node in it
+                            if (clusterService.state().nodes().getMasterNode() != null
+                                && clusterService.state().nodes().nodeExists(joinRequest.getSourceNode().getId()) == false) {
+                                // Remove this listener to avoid memory leaks
+                                clusterService.removeListener(clusterStateListener);
+                                ll.onFailure(e);
                             }
```

Comment:
It seems like it'd be simpler to call `clusterStateListener.clusterChanged` directly here, constructing a "fake" ClusterChangedEvent to carry the current cluster state. I think we want the same logic both ways: particularly if the node is for some reason in the current cluster state at this point then we can complete `ll` successfully too.